xdp-dump/xdpdump.c

```diff
@@ -0,0 +1,628 @@
+	bpf_program__set_attach_target(trace_prog_fexit, prog_fd, info.name);
+
+	/* Load the bpf object into memory */
+	err = bpf_object__load(trace_obj);
+	if (err) {
+		char err_msg[STRERR_BUFSIZE];
+
+		if (err == -EPERM) {
+			pr_debug("Permission denied when loading eBPF object; "
+				 "raising rlimit and retrying\n");
+
+			if (!double_rlimit()) {
+				bpf_object__close(trace_obj);
+				goto rlimit_loop;
+			}
+		}
+
+		libbpf_strerror(err, err_msg, sizeof(err_msg));
+		pr_warn("ERROR: Can't load eBPF object: %s(%d)\n",
+			err_msg, err);
+		goto error_exit;
+	}
+
+	/* Attach trace programs only in the direction(s) needed */
+	if (cfg->rx_capture == PROG_IN || cfg->rx_capture == PROG_INOUT) {
+		trace_link_fentry = bpf_program__attach_trace(trace_prog_fentry);
+		err = libbpf_get_error(trace_link_fentry);
+		if (err) {
+			pr_warn("ERROR: Can't attach XDP trace fentry function: %s\n",
+				strerror(-err));
+			goto error_exit;
+		}
+	}
+
+	if (cfg->rx_capture == PROG_OUT || cfg->rx_capture == PROG_INOUT) {
+		trace_link_fexit = bpf_program__attach_trace(trace_prog_fexit);
+		err = libbpf_get_error(trace_link_fexit);
+		if (err) {
+			pr_warn("ERROR: Can't attach XDP trace fexit function: %s\n",
+				strerror(-err));
+			goto error_exit;
+		}
+	}
+
+	/* Figure out the fd for the BPF_MAP_TYPE_PERF_EVENT_ARRAY trace map */
+	perf_map = bpf_object__find_map_by_name(trace_obj, "xdpdump_perf_map");
+	if (!perf_map) {
+		pr_warn("ERROR: Can't find xdpdump_perf_map in trace program!\n");
+		goto error_exit;
+	}
+	perf_map_fd = bpf_map__fd(perf_map);
+	if (perf_map_fd < 0) {
+		pr_warn("ERROR: Can't get xdpdump_perf_map file descriptor: %s\n",
+			strerror(errno));
+		return false;
+	}
+
+        /* Open the pcap handle */
+	if (cfg->pcap_file) {
+		pcap = pcap_open_dead(DLT_EN10MB, cfg->snaplen);
+		if (!pcap) {
+			pr_warn("ERROR: Can't open pcap dead handler!\n");
+			goto error_exit;
+		}
+
+		pcap_dumper = pcap_dump_open(pcap, cfg->pcap_file);
+		if (!pcap_dumper) {
+			pr_warn("ERROR: Can't open pcap file for writing!\n");
+			goto error_exit;
+		}
+	}
+
+	/* Setup perf context */
+	memset(&perf_ctx, 0, sizeof(perf_ctx));
+	perf_ctx.cfg = cfg;
+	perf_ctx.pcap = pcap;
+	perf_ctx.pcap_dumper = pcap_dumper;
+	perf_ctx.epoch_delta = get_epoch_to_uptime_delta();
+
+	/* Setup perf ring buffers */
+	perf_opts.attr = &perf_attr;
+	perf_opts.event_cb = handle_perf_event;
+	perf_opts.ctx = &perf_ctx;
+	perf_buf = perf_buffer__new_raw(perf_map_fd, PERF_MMAP_PAGE_COUNT,
+					&perf_opts);
+
+	/* Loop trough the dumper */
+	while (!exit_xdpdump) {
+		err = perf_buffer__poll(perf_buf, 1000);
+		if (err < 0 && err != -EINTR) {
+			pr_warn("ERROR: Perf buffer polling failed: %s(%d)",
+				strerror(err), err);
+			goto error_exit;
+		}
+	}
+
+	fprintf(stderr, "\n%"PRIu64" packets captured\n",
+		perf_ctx.captured_packets);
+	fprintf(stderr, "%"PRIu64" packets dropped by perf ring\n",
+		perf_ctx.missed_events / 2);
```

Comment:
Why /2?

Reply:
Good catch, no idea why the /2 is still here, guess some leftover debugging :(